src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs

```diff
@@ -0,0 +1,259 @@
+            // at this time and use it for all uses of the same tag helper type.
+            //
+            // We also want to preserve the ordering of the nodes for testability. So insert at the end of any existing
+            // field nodes.
+            var i = 0;
+            while (i < context.Class.Children.Count && context.Class.Children[i] is FieldDeclarationIntermediateNode)
+            {
+                i++;
+            }
+
+            context.Class.Children.Insert(i, new DefaultTagHelperRuntimeIntermediateNode());
+        }
+
+        private void RewriteBody(TagHelperIntermediateNode node)
+        {
+            for (var i = 0; i < node.Children.Count; i++)
+            {
+                if (node.Children[i] is TagHelperBodyIntermediateNode bodyNode)
+                {
+                    // We only expect one body node.
+                    node.Children[i] = new DefaultTagHelperBodyIntermediateNode(bodyNode);
+                    break;
+                }
+            }
+        }
+
+        private void AddExecute(TagHelperIntermediateNode node)
+        {
+            // Execute the tag helpers at the end, before leaving scope.
+            node.Children.Add(new DefaultTagHelperExecuteIntermediateNode());
+        }
+
+        private void RewriteHtmlAttributes(TagHelperIntermediateNode node)
+        {
+            // We.need to rewrite each html attribute, so that it will get added to the execution context.
+            for (var i = 0; i < node.Children.Count; i++)
+            {
+                if (node.Children[i] is TagHelperHtmlAttributeIntermediateNode htmlAttributeNode)
+                {
+                    node.Children[i] = new DefaultTagHelperHtmlAttributeIntermediateNode(htmlAttributeNode);
+                }
+            }
+        }
+
+        private void RewriteUsage(Context context, TagHelperIntermediateNode node, TagHelperDescriptor tagHelper)
+        {
+            if (!tagHelper.IsDefaultKind())
+            {
+                return;
+            }
+
+            context.Add(tagHelper);
+
+            // First we need to insert a node for the creation of the tag helper, and the hook up to the execution
+            // context. This should come after the body node and any existing create nodes.
+            //
+            // If we're dealing with something totally malformed, then we'll end up just inserting at the end, and that's not
+            // so bad.
+            var i = 0;
+
+            // Find the body node.
+            while (i < node.Children.Count && node.Children[i] is TagHelperBodyIntermediateNode)
+            {
+                i++;
+            }
+            while (i < node.Children.Count && node.Children[i] is DefaultTagHelperBodyIntermediateNode)
+            {
+                i++;
+            }
+
+            // Now find the last create node.
+            while (i < node.Children.Count && node.Children[i] is DefaultTagHelperCreateIntermediateNode)
```

Comment:
I'm being defensive. If I didn't do it here and did it in VCTH you would make the opposite comment